cpp/open3d/t/geometry/TriangleMesh.cpp

```diff
@@ -1061,6 +1061,124 @@ TriangleMesh TriangleMesh::SelectFacesByMask(const core::Tensor &mask) const {
     return result;
 }
 
+/// \brief A helper to compute the vertex and triangle masks based on indices.
+/// Additionally updates tris_cpu to new indices.
+template <typename tri_intT, typename indices_intT>
+static void SBIUpdateMasksAndTrisCPUHelper(const core::Tensor &indices,
+                                           core::Tensor &vertex_mask,
+                                           core::Tensor &tris_mask,
+                                           core::Tensor &tris_cpu) {
+    const int64_t num_tris = tris_cpu.GetLength();
+    const int64_t num_verts = vertex_mask.GetLength();
+
+    // compute the vertices mask
+    tri_intT *vertex_mask_ptr = vertex_mask.GetDataPtr<tri_intT>();
+    const indices_intT *indices_ptr = indices.GetDataPtr<indices_intT>();
+    for (int64_t i = 0; i < indices.GetLength(); ++i) {
+        if (indices_ptr[i] < 0) {
+            utility::LogError(
+                    "[SelectByIndex] indices contains a negative index {}. ",
+                    indices_ptr[i]);
+        }
+        if (indices_ptr[i] >= num_verts) {
+            utility::LogError(
+                    "[SelectByIndex] indices contains index {} out of range. ",
+                    indices_ptr[i]);
+        }
+        vertex_mask_ptr[indices_ptr[i]] = 1;
+    }
+
+    // compute new vertix indices
+    std::vector<tri_intT> prefix_sum(num_verts + 1, 0);
+    utility::InclusivePrefixSum(vertex_mask_ptr, vertex_mask_ptr + num_verts,
+                                &prefix_sum[1]);
+
+    // update the triangles with new indices and build the triangle mask
+    tri_intT *tris_cpu_ptr = tris_cpu.GetDataPtr<tri_intT>();
+    bool *tris_mask_ptr = tris_mask.GetDataPtr<bool>();
+    for (int64_t i = 0; i < num_tris; ++i) {
+        if (vertex_mask_ptr[tris_cpu_ptr[3 * i]] == 1 &&
+            vertex_mask_ptr[tris_cpu_ptr[3 * i + 1]] == 1 &&
+            vertex_mask_ptr[tris_cpu_ptr[3 * i + 2]] == 1) {
+            tris_cpu_ptr[3 * i] = prefix_sum[tris_cpu_ptr[3 * i]];
+            tris_cpu_ptr[3 * i + 1] = prefix_sum[tris_cpu_ptr[3 * i + 1]];
+            tris_cpu_ptr[3 * i + 2] = prefix_sum[tris_cpu_ptr[3 * i + 2]];
+            tris_mask_ptr[i] = true;
+        }
+    }
+}
```

Comment:
In the interest of reducing duplicate code, do you want to factor out the common code from here and `TriangleMesh::SelectFacesByMask()`?

Reply:
I added two static helpers: one to update the triangle indices and one to copy the mesh attributes.